Dockerfile

```diff
@@ -0,0 +1,109 @@
+RUN apt-get update -qq
+RUN apt-get install -qq -o=Dpkg::Use-Pty=0 \
+    git gcc curl attr libgit2-dev zip \
+    fish zsh bash bash-completion
+
+# Install kcov for test coverage
+# This doesn’t run coverage over the xtests so it’s less useful for now
+COPY --from=cargo-kcov /usr/local/cargo/bin/cargo-kcov /usr/bin/cargo-kcov
+
+RUN apt-get install -y \
+    cmake g++ pkg-config \
+    libcurl4-openssl-dev libdw-dev binutils-dev libiberty-dev
+
+RUN ln -s $(which python3) /usr/bin/python
+RUN cargo kcov --print-install-kcov-sh | sh
+
+# Create a variety of misc scripts.
+
+RUN ln -sf /vagrant/devtools/dev-run-debug.sh /usr/bin/exa
+RUN ln -sf /vagrant/devtools/dev-run-release.sh /usr/bin/rexa
+
+RUN echo -e "#!/bin/sh\ncargo build --manifest-path /vagrant/Cargo.toml \\$@" > /usr/bin/build-exa
+RUN ln -sf /usr/bin/build-exa /usr/bin/b
+
+RUN echo -e "#!/bin/sh\ncargo test --manifest-path /vagrant/Cargo.toml \\$@ -- --quiet" > /usr/bin/test-exa
+RUN ln -sf /usr/bin/test-exa /usr/bin/t
+
+RUN echo -e "#!/bin/sh\n/vagrant/xtests/run.sh" > /usr/bin/run-xtests
+RUN ln -sf /usr/bin/run-xtests /usr/bin/x
+
+RUN echo -e "#!/bin/sh\nbuild-exa && test-exa && run-xtests" > /usr/bin/compile-exa
+RUN ln -sf /usr/bin/compile-exa /usr/bin/c
+
+ADD --chmod=+x devtools/dev-package-for-linux.sh /vagrant/devtools/dev-package-for-linux.sh
+RUN echo -e "#!/bin/sh\nbash /vagrant/devtools/dev-package-for-linux.sh \\$@" > /usr/bin/package-exa
+RUN echo -e "#!/bin/sh\ncat /etc/motd" > /usr/bin/halp
+
+# Configure the welcoming text that gets shown:
+
+# Capture the help text so it gets displayed first
+RUN rm -f /etc/update-motd.d/*
+ADD --chmod=+x devtools/dev-help.sh /vagrant/devtools/dev-help.sh
+RUN bash /vagrant/devtools/dev-help.sh > /etc/motd
+
+# Tell bash to execute a bunch of stuff when a session starts
+RUN echo "source /vagrant/devtools/dev-bash.sh" > ${HOME}/.bash_profile
+
+# Link the completion files so they’re “installed”:
+
+# bash
+RUN ln -s /vagrant/completions/bash/exa /etc/bash_completion.d/exa
+RUN ln -s /vagrant/completions/bash/exa /usr/share/bash-completion/completions/exa
+
+# zsh
+RUN ln -s /vagrant/completions/zsh/_exa /usr/share/zsh/vendor-completions/_exa
+
+# fish
+RUN ln -s /vagrant/completions/fish/exa.fish /usr/share/fish/completions/exa.fish
+
+
+# Copy the source code into the image
+COPY --chmod=+x . /vagrant/
+
+# Make sudo dummy replacement
+# This is needed for some tests that use sudo
+RUN echo -e '#!/bin/sh\n"$@"' > /usr/bin/sudo
+RUN chmod +x /usr/bin/sudo
+
+RUN bash /vagrant/devtools/dev-set-up-environment.sh
+RUN bash /vagrant/devtools/dev-create-test-filesystem.sh
+
+WORKDIR /vagrant
+COPY --from=exa /app/target /vagrant/target
+
+# TODO: remove this once tests don't depend on it
+RUN ln -s /vagrant/* ${HOME}
+
+COPY --from=specsheet /usr/local/cargo/bin/specsheet /usr/bin/specsheet
+COPY --from=cargo-hack /usr/local/cargo/bin/cargo-hack /usr/bin/cargo-hack
+COPY --from=just /usr/bin/just /usr/bin/just
```

Comment:
kconv should be located down here with the rest? I assume the concern was due to the package install portion for kconv and moving the layer(s) down to the end here being problematic when an earlier instruction invalidates the `Dockerfile` cache?

Reply:
You could actually shift these up to after kconv, shouldn't be an issue given what the rest of the stage is doing?

Reply:
> kconv should be located down here with the rest?

Actually it's needed for `RUN cargo kcov --print-install-kcov-sh | sh` [here](https://github.com/AgustinRamiroDiaz/exa/blob/bb31f84f9f24dcacebac8edc4891fc3ac6175a5b/Dockerfile#L51)

> You could actually shift these up to after kconv, shouldn't be an issue given what the rest of the stage is doing?

That's true, do you think it's useful? I don't understand the benefit

Reply:
>I don't understand the benefit

The layers that are more frequently to change should come later. The bulk of this stage is something that could run in a separate script via `RUN`, but for now you've stated you want to keep it in the `Dockerfile`.

You could move these `COPY` lines up before `kconv` for better co-location, but right now no major concern either way.